packages/react-client/src/hooks/internal/device/useDevice.ts

```diff
@@ -0,0 +1,172 @@
+import type { SetStateAction } from "react";
+import { useCallback, useMemo, useState } from "react";
+
+import type { DeviceError, DeviceItem, TrackMiddleware } from "../../../types/public";
+import { parseUserMediaError } from "../../../utils/errors";
+import { getTrackFromStream, stopStream } from "../../../utils/track";
+import { useTrackMiddleware } from "../useTrackMiddleware";
+
+type DeviceManagerProps = {
+  mediaStream: MediaStream | null;
+  setStream: (action: SetStateAction<MediaStream | null>) => void;
+  deviceError: DeviceError | null;
+  setError: (action: SetStateAction<DeviceError | null>) => void;
+  getInitialStream: () => Promise<MediaStream | null>;
+  deviceType: "audio" | "video";
+  allDevicesList: MediaDeviceInfo[];
+  constraints?: MediaTrackConstraints | boolean;
+  saveUsedDevice: (device: MediaDeviceInfo) => void;
+};
+
+export type DeviceManager = {
+  startDevice: (deviceId?: string) => Promise<MediaStreamTrack | null>;
+  stopDevice: () => void;
+  activeDevice: DeviceItem | null;
+  deviceTrack: MediaStreamTrack | null;
+  deviceList: DeviceItem[];
+  deviceEnabled: boolean;
+  enableDevice: () => void;
+  disableDevice: () => void;
+  currentMiddleware: TrackMiddleware;
+  applyMiddleware: (middleware: TrackMiddleware) => MediaStreamTrack | null;
+  deviceError: DeviceError | null;
+};
+
+function getReplaceStreamAction(
+  newStream: MediaStream | null,
+  deviceType: "audio" | "video",
+): SetStateAction<MediaStream | null> {
+  return (oldStream: MediaStream | null) => {
+    if (oldStream) {
+      stopStream(oldStream, deviceType);
+    }
+    return newStream;
+  };
+}
+
+function getStopStreamAction(deviceType: "audio" | "video"): SetStateAction<MediaStream | null> {
+  return (oldStream: MediaStream | null) => {
+    if (oldStream) {
+      stopStream(oldStream, deviceType);
+    }
+    return null;
+  };
+}
+
+async function getDeviceStream(
+  type: "audio" | "video",
+  constraints: MediaTrackConstraints | boolean | undefined,
+  deviceId?: string,
+) {
+  if (constraints === false) {
+    console.warn("Attempted to enable camera, but its disabled by a constraint");
+    return null;
+  }
+  constraints = typeof constraints === "object" ? constraints : {};
+  if (deviceId) {
+    constraints.deviceId = { exact: deviceId };
+  }
+  const stream = await navigator.mediaDevices.getUserMedia({
+    [type]: constraints,
+  });
+  return stream;
+}
+
+export const useDeviceManager = ({
+  mediaStream,
+  getInitialStream,
+  deviceType,
+  allDevicesList,
+  setStream,
+  constraints,
+  saveUsedDevice,
+  deviceError,
+  setError,
+}: DeviceManagerProps): DeviceManager => {
+  const rawTrack = useMemo(() => mediaStream && getTrackFromStream(mediaStream, deviceType), [mediaStream, deviceType]);
+
+  const { processedTrack, applyMiddleware, currentMiddleware } = useTrackMiddleware(rawTrack);
+
+  const currentTrack = useMemo(() => processedTrack ?? rawTrack, [rawTrack, processedTrack]);
```

Comment:
I don't think `useMemo` is required here

Reply:
that's correct